client/test/e2e.test.js

```diff
@@ -0,0 +1,112 @@
+import test from 'tape';
+import puppeteer from 'puppeteer';
+
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+const width = 1920;
+const height = 1080;
+
+function browser(width, height) {
+    return puppeteer.launch({
+    headless: true,
+    args: [`--window-size=${width},${height}`],
+    // slowMo: 20 // For testing
+    });
+}
+
+test('e2e-client-test-valid-actions', async (t) => {
+    // Arrange: Set up the page
+    const b = await browser(width, height);
+    const p = await b.newPage();
+    await p.setViewport({ width, height });
+    await p.goto("http://127.0.0.1:8080/");
+    
+    // Act: Test creating a workspace
+    await p.waitForSelector("#add-workspace");
+    await p.click("#add-workspace")
+    await p.waitForSelector("#workspace-name");
+    await p.focus('#workspace-name')
+    await p.keyboard.type('puppeteer')
+    await p.click("#create-workspace")
+
+    // Assert: Check that the new workspace exists with no tables
+    await p.waitForSelector(".v-list-item");
+    await p.waitForSelector(".v-list-item__title");
+    const workspace_name = await p.evaluate(() => document.querySelector('.v-list-item__title').innerText);
```

Comment:
This evaluation is repeated verbatim in the other test--I think it makes sense to abstract it into a function.

Reply:
Fixed in 7087cd9